### ISSUE TITLE
Python package add_files_to_view override breaks hardlink views when using re-installed packages

### ISSUE THREAD

**Issue by dale-mittleman** [LLNL/spack#37091]
### Steps to reproduce

I created an environment.  I added a single package, a prebuilt python.  I then manually edited the environment spec to switch the view type to hardlink.  When I try to regenerate the view, I run into errors of type FileNotFoundError

```
$ echo $PWD 
/hostname/mittleman/spack/admin/environments/python-test

$ cat spack.yaml
spack:
  specs:
    # Python @3.9.15, previously built
    - /v3vlpbp7qvkygzjnyw7v4qh436vktrtm
  view:
    default:
      root: .spack-env/view
      link_type: hardlink
  concretizer:
    unify: true

$ spack spec /v3vlpbp7qvkygzjnyw7v4qh436vktrtm
python@3.9.15%gcc@8.2.1 cflags="-m64 -O2 -g -std=gnu17 -march=sandybridge -mtune=skylake -m64 -O2 -g -std=gnu17 -march=sandybridge -mtune=skylake" cppflags="-std=gnu17 -std=gnu++17 -std=gnu17 -std=gnu++17" cxxflags="-m64 -O2 -g -std=gnu++17 -march=sandybridge -mtune=skylake -m64 -O2 -g -std=gnu++17 -march=sandybridge -mtune=skylake" fflags="-m64 -O2 -g -std=gnu -march=sandybridge -m64 -O2 -g -std=gnu -march=sandybridge" ldflags="-m64 -m64" +bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib build_system=generic patches=0d98e93,4c24573,f2fd060 arch=linux-sles12-x86_64
... dependencies irrelevant...

$ spack --vd --stacktrace env view regenerate
... output below ...
```

### Error message

```console
$ spack -vd --stacktrace env view regenerate
hostname/mittleman/spack/spack/lib/spack/spack/config.py:1010 ==> [2023-04-21-16:35:09.174586] Reading config from file /hostname/mittleman/spack/spack/etc/spack/defaults/config.yaml
hostname/mittleman/spack/spack/lib/spack/spack/config.py:1010 ==> [2023-04-21-16:35:09.198440] Reading config from file /hostname/mittleman/spack/admin/config/spack/config.yaml
hostname/mittleman/spack/spack/lib/spack/spack/config.py:1010 ==> [2023-04-21-16:35:09.206383] Reading config from file /hostname/mittleman/spack/spack/etc/spack/defaults/concretizer.yaml
hostname/mittleman/spack/spack/lib/spack/spack/config.py:1010 ==> [2023-04-21-16:35:09.211930] Reading config from file /hostname/mittleman/spack/admin/config/spack/concretizer.yaml
hostname/mittleman/spack/spack/lib/spack/spack/config.py:1010 ==> [2023-04-21-16:35:09.220205] Reading config from file /hostname/mittleman/spack/admin/environments/python-test/spack.yaml
hostname/mittleman/spack/spack/lib/spack/spack/config.py:1010 ==> [2023-04-21-16:35:09.224131] Reading config from file /hostname/mittleman/spack/spack/etc/spack/defaults/config.yaml
hostname/mittleman/spack/spack/lib/spack/spack/config.py:1010 ==> [2023-04-21-16:35:09.246830] Reading config from file /hostname/mittleman/spack/admin/config/spack/config.yaml
hostname/mittleman/spack/spack/lib/spack/spack/environment/environment.py:203 ==> [2023-04-21-16:35:09.251034] Using environment 'python-test'
hostname/mittleman/spack/spack/lib/spack/spack/cmd/__init__.py:125 ==> [2023-04-21-16:35:09.258282] Imported env from built-in commands
hostname/mittleman/spack/spack/lib/spack/spack/cmd/__init__.py:125 ==> [2023-04-21-16:35:09.263328] Imported env from built-in commands
hostname/mittleman/spack/spack/lib/spack/spack/database.py:391 ==> [2023-04-21-16:35:09.267192] DATABASE LOCK TIMEOUT: 60s
hostname/mittleman/spack/spack/lib/spack/spack/database.py:397 ==> [2023-04-21-16:35:09.267401] PACKAGE LOCK TIMEOUT: No timeout
hostname/mittleman/spack/spack/lib/spack/spack/environment/environment.py:607 ==> [2023-04-21-16:35:09.380842] Updating view at /hostname/mittleman/spack/admin/environments/python-test/.spack-env/view
hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py:666 ==> [2023-04-21-16:35:09.381901] Warning: Skipping external package: bzip2@1.0.6%gcc@8.2.1~debug~pic+shared build_system=generic arch=linux-sles12-x86_64/eyqhmmm
hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py:666 ==> [2023-04-21-16:35:09.382362] Warning: Skipping external package: expat@2.1.0%gcc@8.2.1+libbsd build_system=autotools arch=linux-sles12-x86_64/olbnf3x
hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py:666 ==> [2023-04-21-16:35:09.382803] Warning: Skipping external package: gdbm@1.10%gcc@8.2.1 build_system=autotools arch=linux-sles12-x86_64/53axzaq
hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py:666 ==> [2023-04-21-16:35:09.383245] Warning: Skipping external package: libxml2@2.9.4%gcc@8.2.1~python build_system=autotools arch=linux-sles12-x86_64/x26lsx6
hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py:666 ==> [2023-04-21-16:35:09.383691] Warning: Skipping external package: ncurses@5.9%gcc@8.2.1~symlinks+termlib abi=none build_system=autotools patches=f84b270 arch=linux-sles12-x86_64/x3rdi2k
hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py:666 ==> [2023-04-21-16:35:09.384145] Warning: Skipping external package: zlib@1.2.8%gcc@8.2.1+optimize+pic+shared build_system=makefile arch=linux-sles12-x86_64/puttl5i
hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py:666 ==> [2023-04-21-16:35:09.384586] Warning: Skipping external package: xz@5.0.5%gcc@8.2.1~pic build_system=autotools libs=shared,static arch=linux-sles12-x86_64/vflzfry
hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py:666 ==> [2023-04-21-16:35:09.385021] Warning: Skipping external package: libffi@5.3.1%gcc@8.2.1 build_system=autotools arch=linux-sles12-x86_64/grcdzn7
hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py:666 ==> [2023-04-21-16:35:09.385453] Warning: Skipping external package: libuuid@2.29.2%gcc@8.2.1 build_system=autotools arch=linux-sles12-x86_64/bpbmab7
hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py:666 ==> [2023-04-21-16:35:09.385902] Warning: Skipping external package: openssl@1.0.2p%gcc@8.2.1~docs~shared build_system=generic certs=mozilla arch=linux-sles12-x86_64/yjyltdn
hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py:666 ==> [2023-04-21-16:35:09.386336] Warning: Skipping external package: readline@6.3%gcc@8.2.1 build_system=autotools patches=6633385 arch=linux-sles12-x86_64/3uczhsf
hostname/mittleman/spack/spack/lib/spack/spack/config.py:1010 ==> [2023-04-21-16:35:09.387372] Reading config from file /hostname/mittleman/spack/spack/etc/spack/defaults/repos.yaml
hostname/mittleman/spack/spack/lib/spack/spack/config.py:1010 ==> [2023-04-21-16:35:09.389556] Reading config from file /hostname/mittleman/spack/admin/config/spack/repos.yaml
hostname/mittleman/spack/spack/lib/spack/spack/config.py:1010 ==> [2023-04-21-16:35:09.476213] Reading config from file /hostname/mittleman/spack/spack/etc/spack/defaults/packages.yaml
hostname/mittleman/spack/spack/lib/spack/spack/config.py:1010 ==> [2023-04-21-16:35:09.511505] Reading config from file /hostname/mittleman/spack/admin/config/spack/packages.yaml
hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py:696 ==> [2023-04-21-16:35:09.820139] 1 file conflicts
hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py:700 ==> [2023-04-21-16:35:09.820340] Creating 468 dirs and 9549 links
Traceback (most recent call last):
  File "/hostname/mittleman/spack/spack/bin/spack", line 54, in <module>
    sys.exit(main())
  File "/hostname/mittleman/spack/spack/lib/spack/spack_installable/main.py", line 37, in main
    sys.exit(spack.main.main(argv))
  File "/hostname/mittleman/spack/spack/lib/spack/spack/main.py", line 1004, in main
    return _main(argv)
  File "/hostname/mittleman/spack/spack/lib/spack/spack/main.py", line 959, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/hostname/mittleman/spack/spack/lib/spack/spack/main.py", line 987, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/hostname/mittleman/spack/spack/lib/spack/spack/main.py", line 636, in _invoke_command
    return_val = command(parser, args)
  File "/hostname/mittleman/spack/spack/lib/spack/spack/cmd/env.py", line 837, in env
    action(args)
  File "/hostname/mittleman/spack/spack/lib/spack/spack/cmd/env.py", line 436, in env_view
    env.regenerate_views()
  File "/hostname/mittleman/spack/spack/lib/spack/spack/environment/environment.py", line 1563, in regenerate_views
    view.regenerate(self.concrete_roots())
  File "/hostname/mittleman/spack/spack/lib/spack/spack/environment/environment.py", line 617, in regenerate
    view.add_specs(*specs, with_dependencies=False)
  File "/hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py", line 715, in add_specs
    spec.package.add_files_to_view(self, merge_map, skip_if_exists=False)
  File "/hostname/mittleman/spack/spack/var/spack/repos/builtin/packages/python/package.py", line 1189, in add_files_to_view

  File "/hostname/mittleman/spack/spack/lib/spack/spack/filesystem_view.py", line 57, in view_hardlink
    os.link(src, dst)
FileNotFoundError: [Errno 2] No such file or directory: '/hostname/mittleman/spack/admin/environments/python-test/.spack-env/._view/hsjbxhp5rk5jiopkgvhspoxd6vwq22cc/bin/2to3-3.9' -> '/hostname/mittleman/spack/admin/environments/python-test/.spack-env/._view/hsjbxhp5rk5jiopkgvhspoxd6vwq22cc/bin/2to3'

```

### Information on your system

* **Spack:** 0.20.0.dev0 (4d1100104681db1e0a85e7300efc66ea3c055399)
* **Python:** 3.9.5
* **Platform:** linux-sles12-sandybridge
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by dale-mittleman** [comment]
Some additional detail....

As implied by the issue name, I think the error is a result of the add_files_to_view override in the python package definition, specifically the code dealing with symlinks in the package's bin folder:

https://github.com/spack/spack/blob/d8451b0c3f9ded4c7703fdefb102371138a4189c/var/spack/repos/builtin/packages/python/package.py#L1174-L1188

The merge_map is generated by traversing package trees using os.listdir or os.scandir (depending on which entrypoint into the view system you're using), both of which provide lists of files in no guaranteed order.  This means that links can come before non-links.  For most other package installs, this isn't an issue, because the files in the view are linked directly to the install root, meaning it doesn't matter in what order the merge_map files are processed.  In the case of the python package, because of the way this override re-writes the link targets, order does matter, but only in the case where the view is hardlinked: if you try to hardlink to a file that doesn't exist, an exception is thrown.  

Though I don't have a concrete example, this might also hold for some python extensions, as the same type of logic is applied here: 

https://github.com/spack/spack/blob/d8451b0c3f9ded4c7703fdefb102371138a4189c/lib/spack/spack/build_systems/python.py#L137-L140

I've been looking at this code for a while now, but I can't understand what benefit it provides over just linking back to the install root.  If someone can help me appreciate what the linked code accomplishes, I can try to put together a PR to address this issue

**Comment by dale-mittleman** [comment]
Related to https://github.com/spack/spack/pull/12575

**Comment by alalazo** [comment]
Related to #38580 (we have issues with hash only in spack.yaml)

**Comment by haampie** [comment]
Closed by #42601 (and #42641)
